firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
+def load_c_code(code, name, **kwargs):
+    from pyop2.compilation import load
+    from pyop2.utils import get_petsc_dir
+    cppargs = ["-I%s/include" % d for d in get_petsc_dir()]
+    ldargs = (["-L%s/lib" % d for d in get_petsc_dir()]
+              + ["-Wl,-rpath,%s/lib" % d for d in get_petsc_dir()]
+              + ["-lpetsc", "-lm"])
+    funptr = load(code, "c", name,
+                  cppargs=cppargs, ldargs=ldargs,
+                  **kwargs)
+
+    def get_pointer(obj):
+        if isinstance(obj, (PETSc.Mat, PETSc.Vec)):
+            return obj.handle
+        elif isinstance(obj, numpy.ndarray):
+            return obj.ctypes.data
+        return obj
+
+    @PETSc.Log.EventDecorator(name)
+    def wrapper(*args):
+        return funptr(*list(map(get_pointer, args)))
+    return wrapper
+
+
+def load_assemble_csr(comm, triu=False):
+    if triu:
+        name = "setSubMatCSR_SBAIJ"
+        select_cols = "icol < irow ? -1: icol"
+    else:
+        name = "setSubMatCSR_AIJ"
+        select_cols = "icol"
+    code = f"""
+#include <petsc.h>
+
+PetscErrorCode {name}(Mat A,
+                      Mat B,
+                      PetscInt *rindices,
+                      PetscInt *cindices,
+                      InsertMode addv)
+{{
+    PetscInt ncols, irow, icol;
+    PetscInt *cols, *indices;
+    PetscScalar *vals;
+
+    PetscInt m, n;
+    PetscErrorCode ierr;
+    PetscFunctionBeginUser;
+    MatGetSize(B, &m, NULL);
+
+    n = 0;
+    for (PetscInt i = 0; i < m; i++) {{
+        ierr = MatGetRow(B, i, &ncols, NULL, NULL);CHKERRQ(ierr);
+        n = ncols > n ? ncols : n;
+        ierr = MatRestoreRow(B, i, &ncols, NULL, NULL);CHKERRQ(ierr);
+    }}
+    PetscMalloc1(n, &indices);
+    for (PetscInt i = 0; i < m; i++) {{
+        ierr = MatGetRow(B, i, &ncols, &cols, &vals);CHKERRQ(ierr);
+        irow = rindices[i];
+        for (PetscInt j = 0; j < ncols; j++) {{
+            icol = cindices[cols[j]];
+            indices[j] = {select_cols};
+        }}
+        ierr = MatSetValues(A, 1, &irow, ncols, indices, vals, addv);CHKERRQ(ierr);
+        ierr = MatRestoreRow(B, i, &ncols, &cols, &vals);CHKERRQ(ierr);
+    }}
+    PetscFree(indices);
+    PetscFunctionReturn(0);
+}}
+"""
+    argtypes = [ctypes.c_voidp, ctypes.c_voidp,
+                ctypes.c_voidp, ctypes.c_voidp, ctypes.c_int]
+    return load_c_code(code, name, comm=comm, argtypes=argtypes,
+                       restype=ctypes.c_int)
+
+
+def petsc_sparse(A_numpy, rtol=1E-10):
+    Amax = max(A_numpy.min(), A_numpy.max(), key=abs)
+    atol = rtol*Amax
+    nnz = numpy.count_nonzero(abs(A_numpy) > atol, axis=1).astype(PETSc.IntType)
+    A = PETSc.Mat().createAIJ(A_numpy.shape, nnz=(nnz, 0), comm=PETSc.COMM_SELF)
+    for row, Arow in enumerate(A_numpy):
+        cols = numpy.argwhere(abs(Arow) > atol).astype(PETSc.IntType).flat
+        A.setValues(row, cols, Arow[cols], PETSc.InsertMode.INSERT)
+    A.assemble()
+    return A
+
+
+def block_mat(A_blocks):
```

Comment:
Comment "A_blocks is an array of arrays defining a block matrix of PETSc Mats. Returns the corresponding overall PETSc Mat. We make an AIJ instead of a nest because ..."